core/src/main/scala/BinaryTree.scala

```diff
@@ -0,0 +1,251 @@
+        case EQ                     =>  this.asInstanceOf[Branch[A]]
+        case GT                     =>  Branch(Some(a), l, r.add(x)(order))
+      }
+    }
+
+    balance(insert(x, order)).asInstanceOf[Branch[A]]
+  }
+
+  def +(x: A)(implicit order: Order[A]): BinaryTree[A] = add(x)
+
+  def remove(x: A)(implicit order: Order[A]): BinaryTree[A] = {
+    def del(x:A)(implicit order: Order[A]): BinaryTree[A] = this match {
+      case BTNil() => BinaryTree.empty
+      case Branch(Some(a), l, r, _, _) => order.compare(x, a) match {
+        case LT => Branch(Some(a), l.remove(x), r)
+        case GT => Branch(Some(a), l, r.remove(x))
+        case EQ => (l, r) match {
+          case (BTNil(), BTNil()) => BinaryTree.empty
+          case (BTNil(), x) => x
+          case (x, BTNil()) => x
+          case (x, y) => {
+            val min = y.min
+
+            min match {
+              case Some(a) => Branch(min, x, y.remove(a))
+              case None() => x //<- this should never happen
+            }
+          }
+        }
+      }
+    }
+
+    balance(del(x))
+  }
+
+  def join(another: BinaryTree[A])(implicit order: Order[A]) = {
+
+    @tailrec def build(sub: BinaryTree[A], xs: List[A]): BinaryTree[A] = xs match {
+      case El()                 =>  sub
+      case Nel(h, t)            =>  build(sub + h, t)
+    }
+
+    another match {
+      case BTNil()              =>  this
+      case _                    =>  build(this, another.toLst())
+    }
+  }
+
+  def ++(another: BinaryTree[A])(implicit order: Order[A]) = join(another)
+
+  private def balance(tree: BinaryTree[A]): BinaryTree[A] = {
+    var t = tree.update
+
+    if (t.balanceFactor > 1) {
+      if (t.right.balanceFactor <= 0) {
+        t = Branch(t.value, t.left, rotate(t.right, 1), t.size, t.height)
+      }
+      t = rotate(t, -1)
+    } else {
+      if (t.balanceFactor < -1)
+      {
+        if (t.left.balanceFactor >= 0){
+          t = Branch(t.value, rotate(t.left, -1), t.right, t.size, t.height)
+        }
+        t = rotate(t, 1)
+      }
+    }
+
+    return t
+  }
+
+  private def rotate(x: BinaryTree[A], direction: Int): BinaryTree[A] = {
+    if (direction < 0) {  // left
+
+      if (x == BTNil() || x.right == BTNil()) return x
+
+      val y = x.right
+      val a = Branch(x.value, x.left, BTNil(), x.size, x.height).update()
+
+      val b = Branch(y.value, a, y.right, y.size, y.height).update()
+
+      return b
+    }
+    else { //right
+      if (x == BTNil() || x.left == BTNil()) return x
+
+      val y = x.left
+      val a = Branch(x.value, BTNil(), x.right, x.size, x.height).update()
+
+      val b = Branch(y.value, y.left, a, y.size, y.height).update()
+
+      return b
+    }
+  }
+
+}
+
+
+
+case class Branch[A](value: Option[A],
+                     left: BinaryTree[A],
+                     right: BinaryTree[A],
+                     private [dogs] val size: Int = 0,
```

Comment:
I cannot remove this field from here since sometimes they need to be initialized with a special values. Anyways they are private to the package.

We can recalculate them all the time, but during balancing they will be needed it a lot of time since balancing is recursive the same value will be calculated a lot of time. I rather have them calculated when going up in the tree and have them ready for next time